File: test/test_binary_partition_tree.cpp
/***************************************************************************
* Copyright ESIEE Paris (2018)                                             *
*                                                                          *
* Contributor(s) : Benjamin Perret                                         *
*                                                                          *
* Distributed under the terms of the CECILL-B License.                     *
*                                                                          *
* The full license is in the file LICENSE, distributed with this software. *
****************************************************************************/

#include <boost/test/unit_test.hpp>
#include "test_utils.hpp"
#include "higra/hierarchy/binary_partition_tree.hpp"
#include "higra/algo/graph_image.hpp"
#include "xtensor/xrandom.hpp"
#include "higra/hierarchy/hierarchy_core.hpp"
#include "higra/algo/tree.hpp"


using namespace hg;
using namespace std;

BOOST_AUTO_TEST_SUITE(test_binary_partition_tree);

    BOOST_AUTO_TEST_CASE(test_single_linkage_clustering_simple) {
        auto graph = get_4_adjacency_graph({3, 3});
        array_1d<double> edge_weights({1, 9, 6, 7, 5, 8, 12, 4, 10, 11, 2, 3});
        auto res = hg::binary_partition_tree(graph,
                                             edge_weights,
                                             hg::make_binary_partition_tree_min_linkage(
                                                     edge_weights));
        auto &tree = res.first;
        auto &levels = res.second;

        array_1d<index_t> expected_parents({9, 9, 13, 15, 12, 12, 10, 10, 11, 14, 11, 16, 13, 14, 15, 16, 16});
        array_1d<double> expected_levels({0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 8, 10});

        BOOST_CHECK(expected_parents == tree.parents());
        BOOST_CHECK(expected_levels == levels);
    }

    BOOST_AUTO_TEST_CASE(test_single_linkage_clustering_hard) {
        long size = 100;
        auto graph = get_4_adjacency_graph({size, size});
        array_1d<double> edge_weights = xt::random::rand<double>({num_edges(graph)});
        auto res = hg::binary_partition_tree(graph,
                                             edge_weights,
                                             hg::make_binary_partition_tree_min_linkage(edge_weights));
        auto &tree = res.first;


        auto res2 = hg::bptCanonical(graph, edge_weights);
        auto &tree2 = std::get<0>(res2);

        BOOST_CHECK(hg::testTreeIsomorphism(tree, tree2));
    }


BOOST_AUTO_TEST_SUITE_END();

File: python/test/test_binary_partition_tree.py
############################################################################
# Copyright ESIEE Paris (2018)                                             #
#                                                                          #
# Contributor(s) : Benjamin Perret                                         #
#                                                                          #
# Distributed under the terms of the CECILL-B License.                     #
#                                                                          #
# The full license is in the file LICENSE, distributed with this software. #
############################################################################

import unittest
import higra as hg
import numpy as np


class TestBinaryPartitionTree(unittest.TestCase):

    def test_binary_partition_tree_complete_linkage(self):
        graph = hg.get_4_adjacency_graph((3, 3))
        edge_weights = np.asarray((1, 8, 2, 10, 15, 3, 11, 4, 12, 13, 5, 6), np.float32)
        res = hg._binary_partition_tree_complete_linkage(graph, edge_weights)
        tree = res.tree()
        levels = res.node_altitude()

        expected_parents = np.asarray((9, 9, 10, 11, 11, 12, 13, 13, 14, 10, 16, 12, 15, 14, 15, 16, 16), np.uint32)
        expected_levels = np.asarray((0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 13, 15), np.float32)

        self.assertTrue(np.all(expected_parents == tree.parents()))
        self.assertTrue(np.all(expected_levels == levels))

    def test_binary_partition_tree_average_linkage(self):
        graph = hg.get_4_adjacency_graph((3, 3))
        edge_values = np.asarray((1, 7, 2, 10, 16, 3, 11, 4, 12, 14, 5, 6), np.float32)
        edge_weights = np.asarray((7, 1, 7, 3, 2, 8, 2, 2, 2, 1, 5, 9), np.float32)
        res = hg._binary_partition_tree_average_linkage(graph, edge_values, edge_weights)
        tree = res.tree()
        levels = res.node_altitude()

        expected_parents = np.asarray((9, 9, 10, 11, 11, 12, 13, 13, 14, 10, 15, 12, 15, 14, 16, 16, 16), np.uint32)
        expected_levels = np.asarray((0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 11.5, 12), np.float32)

        self.assertTrue(np.all(expected_parents == tree.parents()))
        self.assertTrue(np.all(expected_levels == levels))

    def test_binary_partition_tree_custom_linkage(self):
        graph = hg.get_4_adjacency_graph((3, 3))
        edge_values = np.asarray((1, 7, 2, 10, 16, 3, 11, 4, 12, 14, 5, 6), np.float32)
        edge_weights = np.asarray((7, 1, 7, 3, 2, 8, 2, 2, 2, 1, 5, 9), np.float32)

        def weighting_function_average_linkage(graph, fusion_edge_index, new_region, merged_region1, merged_region2, new_neighbours):
            for n in new_neighbours:
                if n.num_edges() > 1:
                    new_weight = edge_weights[n.first_edge_index()] + edge_weights[n.second_edge_index()]
                    new_value = (edge_values[n.first_edge_index()] * edge_weights[n.first_edge_index()] \
                        + edge_values[n.second_edge_index()] * edge_weights[n.second_edge_index()]) \
                        / new_weight
                else:
                    new_weight = edge_weights[n.first_edge_index()]
                    new_value = edge_values[n.first_edge_index()]

                n.set_new_edge_weight(new_value)
                edge_values[n.new_edge_index()] = new_value
                edge_weights[n.new_edge_index()] = new_weight

        res = hg._binary_partition_tree(graph, edge_values, weighting_function_average_linkage)
        tree = res.tree()
        levels = res.node_altitude()

        expected_parents = np.asarray((9, 9, 10, 11, 11, 12, 13, 13, 14, 10, 15, 12, 15, 14, 16, 16, 16), np.uint32)
        expected_levels = np.asarray((0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 11.5, 12), np.float32)

        self.assertTrue(np.all(expected_parents == tree.parents()))
        self.assertTrue(np.all(expected_levels == levels))


if __name__ == '__main__':
    unittest.main()
